neureka/hal/neureka_task.c

```diff
@@ -0,0 +1,201 @@
+                                   : NEUREKA_INPUT_CHANNEL_THROUGHPUT_1x1,
+                           .kernel_shape = kernel_shape,
+                           .depthwise = depthwise,
+                           .data = {0}};
+
+  const int flag_mode = kernel_shape == 1 ? NEUREKA_FLAG_MODE_1x1
+                        : depthwise == 1  ? NEUREKA_FLAG_MODE_3x3_DW
+                                          : NEUREKA_FLAG_MODE_3x3;
+
+  task->data.cfg.conf0 |=
+      flag_input_signed << NEUREKA_SHIFT_FLAG_INPUT_SIGNED |
+      NEUREKA_FLAG_NORM_QUANT | quant.function | quant.mode |
+      (quant.shift_amount << 16) |
+      quant.flag_rounding << NEUREKA_SHIFT_ROUNDING | norm.mode |
+      norm.flag_bias << NEUREKA_SHIFT_FLAG_NORM_BIAS |
+      norm.flag_shift << NEUREKA_SHIFT_FLAG_NORM_SHIFT | NEUREKA_FLAG_USE_TCDM |
+      weights_offset_mode | flag_mode | (weights_bits - 1);
+
+  task->data.cfg.weight_offset_factor = weights_offset_factor;
+}
+
+/** neureka_pad_ptr
+ *
+ * Calculate the pointer to the start of the ptr as if
+ * it was the start to the padded data.
+ * Necessary for input pointer when it's padded.
+ */
+inline uint32_t neureka_pad_ptr(uint32_t ptr, const uint32_t width,
+                                const uint32_t channel, const uint8_t bits,
+                                const uint8_t padding_top,
+                                const uint8_t padding_left) {
+  return ptr - (padding_top * width + padding_left) * channel * bits / 8;
+}
+
+inline void neureka_task_set_ptrs(neureka_task_t *task, uint32_t input_ptr,
+                                  uint32_t w_in, uint32_t k_in, uint8_t bits_in,
+                                  uint8_t padding_top, uint8_t padding_left,
+                                  uint32_t output_ptr, uint32_t weights_ptr,
+                                  uint32_t scale_ptr, uint32_t shift_ptr,
+                                  uint32_t bias_ptr) {
+  task->data.infeat_ptr = neureka_pad_ptr(input_ptr, w_in, k_in, bits_in,
+                                          padding_top, padding_left);
+  task->data.outfeat_ptr = output_ptr;
+  task->data.weights_ptr = weights_ptr;
+  task->data.scale_ptr = scale_ptr;
+  task->data.scale_shift_ptr = shift_ptr;
+  task->data.scale_bias_ptr = bias_ptr;
+}
+
+void neureka_task_set_strides(neureka_task_t *task, const uint32_t k_in,
+                              const uint32_t w_in_stride,
+                              const uint32_t k_in_stride,
+                              const uint32_t w_out_stride,
+                              const uint32_t k_out_stride) {
+  const uint32_t num_k_in = divnceil(k_in, task->input_channel_throughput);
+
+  const neureka_stride_t input_stride = {
+      .d0 = k_in_stride,
+      .d1 = k_in_stride * w_in_stride,
+      .d2 = 0 // Unused
+  };
+  task->data.cfg.input_stride = input_stride;
+
+  const neureka_stride_t output_stride = {
+      .d0 = 32, // TODO: should depend on outbytes. Probably 32 / outbytes
```

Comment:
I'd verify this TODO before merging

Reply:
Todo verified. It is in fact not probably `32 / outbytes` but just 32, which is the output bandwidth in bytes. Added a define.